### PR TITLE
WIP: [feature] update react-modal to work with React 16.

### LINKED ISSUE #498: React 16 compatibility
### Summary:

I looked around in the issue tracker but couldn't find an existing issue. As React 16 is right around the corner, maybe it's time to prepare a PR that adds support for it?

It looks like at least the [call to `unstable_renderSubtreeIntoContainer`](https://github.com/reactjs/react-modal/blob/f007aeb7336bc8a47bb6b8c7919ed3a077fbd87d/src/components/Modal.js#L11) should be updated as there's now first class support for portals.

### Steps to reproduce:

 1. Render a modal.
 2. Unmount it.

### Expected behavior:

Shouldn't throw any errors.

### Link to example of issue:

https://codesandbox.io/s/xpl8r853o4

### PR DESCRIPTION
**Pull request by diasbruno**
This update is based on the implementation on [CogniStreamer/react-modal@4434c4a](https://github.com/CogniStreamer/react-modal/commit/4434c4aac57d91a24d8f1b89e85a41fe2bd2023d).

It's still WIP, need to update documentation.

Fixes #498.
Related to #484.

Changes proposed:
- Use the correct way to work with portal in react 16.

Upgrade Path (for changed or removed APIs):
- Update your project to use react 16.

Acceptance Checklist:
- [x] All commits have been squashed to one.
- [x] The commit message follows the guidelines in `CONTRIBUTING.md`.
- [ ] Documentation (README.md) and examples have been updated as needed.
- [x] If this is a code change, a spec testing the functionality has been added.
- [x] If the commit message has [changed] or [removed], there is an upgrade path above.

### CODE REVIEW

**Comment by diasbruno** [comment]
I didn't remove `unmountComponentAtNode(node)` yet, because the render method of `Modal` will have to change to return `null`.

**Comment by bjrmatos** [comment]
is still rendering normally when doing server side rendering? i think that returning `null` was the thing that make possible to not break the component when doing SSR.

**Comment by bjrmatos** [comment]
also doing `this.node = document.createElement('div');` in `componentDidMount` was other of the reasons, moving to the constructor breaks SSR.

**Comment by diasbruno** [comment]
@bjrmatos Thanks. Can you PR these changes to `diasbruno/react-moda:feature/react-16`?

**Comment by catamphetamine** [comment]
Current changes do break server side rendering as @bjrmatos says.
I'd leave `this.node = ...` code where it was, in `componentDidMount` because it's the only place where `document` exists.
And then in `componentDidMount` it would just `this.setState({ node: this.node })` to rerender (or just `this.forceUpdate()`).

File: package.json [hunk 1]
```diff
@@ -62,8 +62,8 @@
     "react-dom-factories": "^1.0.0"
   },
   "peerDependencies": {
-    "react": "^0.14.0 || ^15.0.0",
-    "react-dom": "^0.14.0 || ^15.0.0"
+    "react": "^16.0.0",
+    "react-dom": "^16.0.0"
```
Comment (emilyrohrbough) [inline package.json]: Are you removing compatibility for `^0.14.0 || ^15.0.0`?

**Comment by catamphetamine** [comment]
> @emilyrohrbough Are you removing compatibility for `^0.14.0 || ^15.0.0`?

Yeah, that's also a controversial decision made in the changes.
It should support all Reacts instead, feature-detecting the `ReactDOM.createPortal()` method existence and then behaving the old way in case of older React and behaving the new way in case of a newer React.

File: package.json (hunk 1 above)
Comment (diasbruno) [reply package.json]: Not really, it's just a initial PR :)

**Comment by diasbruno** [comment]
@catamphetamine @emilyrohrbough This PR is available for everyone to participate, give some ideas, etc...It still WIP, as you can see in the title.

**Comment by catamphetamine** [comment]
@diasbruno I took your ideas further and introduced a feature flag so that it's compatible with both old React and new React.
I didn't test the code though.

File: package.json (hunk 1 above)
Comment (diasbruno) [reply package.json]: I'm forcing this temporary just to test what would need to change in regarding to support react 16. If possible, I'll guarantee backwards compatibility with previous version of react.

**Comment by diasbruno** [comment]
@catamphetamine Thanks. You can PR you changes to my branch. So we can keep all in one place.

**Comment by catamphetamine** [comment]
@diasbruno "You may not have permission to access react-modal. Check Preferences to make sure you’re still logged in."

If you want my changes in this branch then you should add me as a collaborator in the settings

**Comment by diasbruno** [comment]
@catamphetamine I'll close this issue in favor of #513.

**Comment by andrepcg** [comment]
Using react-modal@next with Next.js to server side render a react app I'm getting

image

`document is not defined
ReferenceError: document is not defined
    at Modal.render (/Users/aper/DEV/backen/node_modules/react-modal/lib/components/Modal.js:150:21)`

**Comment by diasbruno** [comment]
@andrepcg Please, you can open a issue. 

Probably checking for `typeof window !== undefined` is not enough.

**Comment by catamphetamine** [comment]
@diasbruno Perhaps you should check `document` instead of `window` because it makes more sense since you're not using `window`

**Comment by catamphetamine** [comment]
@diasbruno MAAAAN
WTF
```js
typeof window !== undefined
```
No no

```js
typeof window !== 'undefined'
```

**Comment by catamphetamine** [comment]
@andrepcg Don't open an issue, the bug is found

**Comment by diasbruno** [comment]
@catamphetamine Oh my. This kind of thing happen at 3am.

**Comment by diasbruno** [comment]
I'll fix it and release a new version. Thanks, @catamphetamine.

**Comment by diasbruno** [comment]
@catamphetamine @andrepcg #516 should be a safer check for `canUseDOM`. Can you review it, please?

**Pull request closed without merging**